content/blog/waddlesplash/2023-04-24_condition_variables.md

```diff
@@ -0,0 +1,416 @@
+
+### The new implementation
+
+My fundamental change in approach with the new implementation was to have the `Entry` be not protected by any lock, but have all its member variables be atomically updated, whether by itself or by the `Variable`. Those same variables would then also serve to indicate, depending on their precise state, what state the variable itself was in. This still took some trial and error to get entirely correct; but the final version of the algorithm has run without causing any problems (that we know of) for over a year, now.
+
+So, let’s look at the implementation, one piece at a time.
+
+#### Member variables
+
+The member variables of `ConditionVariable` are as follows (some convenience ones entirely unrelated to synchronization are omitted):
+```cpp
+			typedef DoublyLinkedList<ConditionVariableEntry> EntryList;
+
+			spinlock			fLock;
+			EntryList			fEntries;
+			int32				fEntriesCount;
+```
+We have a `spinlock` protecting the linked-list of entries, and then an `int32 fEntriesCount` specifying the number of entries in the list. This last value is only *partially* protected by `fLock`, and is atomically updated: as we will see later on, it is critically important for synchronizing list removals.
+
+Now, the member variables of `ConditionVariableEntry`:
+```cpp
+			ConditionVariable*	fVariable;
+			Thread*				fThread;
+			status_t			fWaitStatus;
+```
+The first is, of course, a pointer to the `Variable` this `Entry` has been added to (or `NULL` if it has not been added to one; or if it was added, but has since been removed.) `fVariable` being non-`NULL` *always* implies that the `Variable` in question also has this `Entry` in its `fEntries` list; the two are never changed except in tandem.
+
+The second is a pointer to the kernel `Thread` structure, which indicates the thread that this `Entry` belongs to, which will wait on it. The `Variable` uses this to wake up the thread in question, but it is also used to signal (by either the `Entry` itself or the `Variable`) that the `Entry` is in the process of being removed from the `Variable`. One principle that is important to note is that whichever party unsets `fThread` is also the same party responsible for decrementing `fEntriesCount`, as we will see later on.
+
+Finally, we have `fWaitStatus`. This value is either positive (if the `Entry` has been `Add`ed somewhere, but not yet notified and woken up), or `<= 0` (once it has been.) When `Wait` is invoked, it will either return this value or a timeout-related error code, depending on what happens. It has some relevance for synchronization, but not very much, as we will (once again) see later on.
+
+#### `Add`ing `Entry` to a `Variable`
+
+This is the most straightforward action of all: the `Variable`’s `fLock` is acquired, and then the following actions are taken inside the `Entry`:
+```
+inline void
+ConditionVariableEntry::_AddToLockedVariable(ConditionVariable* variable)
+{
+	ASSERT(fVariable == NULL);
```

Comment:
all other assertions later on omitted for brevity's sake, this one probably can be too.